src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -45,5 +32,213 @@ private static IEnumerable<TSource> PrependIterator<TSource>(IEnumerable<TSource
             yield return element;
             foreach (TSource e1 in source) yield return e1;
         }
+
+        public static IEnumerable<TSource> Concat<TSource>(this IEnumerable<TSource> first, IEnumerable<TSource> second)
+        {
+            if (first == null) throw Error.ArgumentNull("first");
+            if (second == null) throw Error.ArgumentNull("second");
+
+            var concatFirst = first as ConcatIterator<TSource>;
+            return concatFirst != null ?
+                concatFirst.Append(second) :
+                new Concat2Iterator<TSource>(first, second);
+        }
+
+        private sealed class Concat2Iterator<TSource> : ConcatIterator<TSource>
+        {
+            private readonly IEnumerable<TSource> _first;
+            private readonly IEnumerable<TSource> _second;
+
+            internal Concat2Iterator(IEnumerable<TSource> first, IEnumerable<TSource> second)
+            {
+                Debug.Assert(first != null && second != null);
+                _first = first;
+                _second = second;
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new Concat2Iterator<TSource>(_first, _second);
+            }
+
+            internal override ConcatIterator<TSource> Append(IEnumerable<TSource> next)
+            {
+                return new Concat3Iterator<TSource>(_first, _second, next);
+            }
+
+            internal override IEnumerable<TSource> GetEnumerable(int index)
+            {
+                switch (index)
+                {
+                    case 0: return _first;
+                    case 1: return _second;
+                    default: return null;
+                }
+            }
+        }
+
+        private sealed class Concat3Iterator<TSource> : ConcatIterator<TSource>
```

Comment:
I could.  Though I actually considered deleting the 3 case even.  Once I switched the N version to using a chain rather than an array, there's no longer a "steep cliff".